hibernate-core/src/main/java/org/hibernate/query/sqm/internal/SqmUtil.java

```diff
@@ -127,27 +128,62 @@ public static IllegalQueryOperationException expectingNonSelect(SqmStatement<?>
 		);
 	}
 
+	/**
+	 * Utility that returns {@code true} if the specified {@link SqmPath sqmPath} has to be
+	 * dereferenced using the target table mapping, and does not support fk optimization.
+	 * <p>
+	 * This is the case when the path is used in both the {@linkplain Clause#GROUP group by} clause
+	 * and a clause which {@linkplain #isClauseDependantOnGroupBy depends on it} or when the left
+	 * hand side of the path is a {@linkplain #isNonOptimizableJoin non-optimizable join}.
+	 */
 	public static boolean needsTargetTableMapping(
 			SqmPath<?> sqmPath,
 			ModelPartContainer modelPartContainer,
 			SqmToSqlAstConverter sqlAstCreationState) {
-		final Clause currentClause = sqlAstCreationState.getCurrentClauseStack().getCurrent();
-		return ( currentClause == Clause.GROUP || currentClause == Clause.SELECT || currentClause == Clause.ORDER || currentClause == Clause.HAVING )
-				&& modelPartContainer.getPartMappingType() != modelPartContainer
+		return modelPartContainer.getPartMappingType() != modelPartContainer
```

Comment:
The change of order in which checks are made has the potential to execute more code than before. Also, I think the explanation for `isClauseDependantOnGroupBy` is not quite right.

Maybe this is a good chance to step back a bit and question how this check came to be. The check was introduced with https://hibernate.atlassian.net/browse/HHH-16189 and the aim was to ensure we use the FK target for paths, when the path also appears in the group by clause, because otherwise the functional dependency analysis of the DB would fail.

But does it make sense to make any of this dependent on the clause in which the path appears or whether the path also appears in the group by clause? Shouldn't we rather always use the FK target if the LHS is a `SqmFrom` and reduce the whole check to the following?

```java
return sqmPath.getLhs() instanceof SqmFrom<?, ?>;
```

I'm pretty sure I might miss something, but I think it's worth trying this out to understand better what optimizations this enables, as well as document these optimizations in here if they are existent.

Reply:
Yeah I know that more code will be executed, but the group-by was there only for those clauses so I wanted to keep it and needed to generalize it for joins.

We can certainly try what you're suggesting, but that would result in basically never using the foreign key side when dereferencing associations - I guess we should keep the group-by or non optimizable join too at least.

Reply:
>We can certainly try what you're suggesting, but that would result in basically never using the foreign key side when dereferencing associations - I guess we should keep the group-by or non optimizable join too at least.

Never use if the LHS is an explicit from clause i.e. `from MyEntity e join e.association a where a.id = 1` would then always generate the join, even if it weren't absolutely necessary. Not sure if that is such a bad thing, but I wanted to suggest the idea of simplifying this code, given that the only optimization that it enables is probably the join avoidance for a query like `from MyEntity e join e.association a where a.id = 1`.

Maybe you can tell about other optimizations this enables when taking my suggestion for a test drive.

Reply:
I've applied your suggestion, though I had to keep the `modelPartContainer` checks to avoid CCEs after. Explicitly joined paths will now always be dereferenced using the target table's mapping thus triggering join creation in the query. I had to change a couple tests which expected fk optimization to occur, but other than that there seems to be no major disruptions.

I suppose since the joins are explicit in HQL it makes sense to always create them in SQL, I leave to you any consideration on the possible performance impacts of this change.